// select2-vue-component/dist/auto-complete.js
import { defineComponent, nextTick } from 'vue';
import * as common from 'select2-component';
export * from 'select2-component';
import { autoCompleteTemplateHtml } from './variables';
/**
 * @public
 */
export var AutoComplete = defineComponent({
    render: autoCompleteTemplateHtml,
    props: {
        data: {
            type: Array,
            required: true,
        },
        value: [String, Number, Boolean, Array],
    },
    data: function () {
        return {
            option: null,
            hoveringValue: null,
            isOpen: false,
            focusoutTimer: undefined,
            lastScrollTopIndex: 0,
            searchInputElement: undefined,
            resultsElement: undefined,
        };
    },
    computed: {
        dropdownStyle: function () {
            return common.getDropdownStyle(this.isOpen && this.data.length > 0);
        },
        filteredData: function () {
            var result = this.data;
            if (common.valueIsNotInFilteredData(result, this.hoveringValue)) {
                this.hoveringValue = common.getFirstAvailableOption(result);
                if (this.resultsElement) {
                    var lastScrollTopIndex = common.getLastScrollTopIndex(this.hoveringValue, this.resultsElement, result, this.lastScrollTopIndex);
                    if (lastScrollTopIndex !== null) {
                        this.lastScrollTopIndex = lastScrollTopIndex;
                    }
                }
            }
            return result;
        },
        containerStyle: function () {
            return common.getContainerStyle(false, this.isOpen && this.data.length > 0);
        }
    },
    beforeMount: function () {
        var option = common.getOptionsByValue(this.data, this.value, false);
        if (option !== null) {
            this.option = option;
        }
        if (!Array.isArray(option)) {
            this.hoveringValue = this.value;
        }
    },
    mounted: function () {
        this.searchInputElement = this.$refs.searchInput;
        this.resultsElement = this.$refs.results;
    },
    methods: {
        getOptionStyle: function (option) {
            return common.getOptionStyle(option, this.hoveringValue);
        },
        mouseenter: function (option) {
            if (!option.disabled) {
                this.hoveringValue = option.value;
            }
        },
        click: function (option) {
            if (!option.disabled) {
                this.select(option);
            }
            if (this.focusoutTimer) {
                clearTimeout(this.focusoutTimer);
            }
        },
        toggleOpenAndClose: function () {
            var _this = this;
            this.isOpen = !this.isOpen;
            if (this.isOpen) {
                nextTick(function () {
                    _this.focusSearchboxOrResultsElement();
                    if (_this.resultsElement) {
                        var lastScrollTopIndex = common.getLastScrollTopIndex(_this.hoveringValue, _this.resultsElement, _this.data, _this.lastScrollTopIndex);
                        if (lastScrollTopIndex !== null) {
                            _this.lastScrollTopIndex = lastScrollTopIndex;
                        }
                    }
                });
            }
            if (this.focusoutTimer) {
                clearTimeout(this.focusoutTimer);
            }
        },
        focusout: function () {
            var _this = this;
            this.focusoutTimer = setTimeout(function () {
                _this.isOpen = false;
                _this.focusoutTimer = undefined;
            }, common.timeout);
        },
        select: function (option) {
            if (option !== null) {
                this.option = option;
                this.isOpen = false;
            }
            this.$emit('select', this.option.value);
            this.$emit('update', this.option.value);
        },
        keyDown: function (e) {
            this.$emit('keydown', e);
            if (e.keyCode === 40) {
                this.moveDown();
                e.preventDefault();
            }
            else if (e.keyCode === 38) {
                this.moveUp();
                e.preventDefault();
            }
            else if (e.keyCode === 13) {
                this.selectByEnter();
                e.preventDefault();
            }
        },
        keyUp: function (e) {
            this.$emit('keyup', e);
        },
        keyPress: function (e) {
            this.$emit('keypress', e);
        },
        onChange: function (e) {
            this.$emit('search', e.target.value);
            this.$emit('update', e.target.value);
        },
        isSelected: function (option) {
            return common.isSelected(this.option, option, false);
        },
        isDisabled: function (option) {
            return option.disabled ? 'true' : 'false';
        },
        removeSelection: function (e, option) {
            var _this = this;
            common.removeSelection(this.option, option);
            this.$emit('update', this.option.map(function (op) { return op.value; }));
            e.preventDefault();
            e.stopPropagation();
            if (this.isOpen) {
                nextTick(function () {
                    _this.focusSearchboxOrResultsElement();
                });
            }
            if (this.focusoutTimer) {
                clearTimeout(this.focusoutTimer);
            }
        },
        focusSearchboxOrResultsElement: function () {
            if (this.searchInputElement) {
                this.searchInputElement.focus();
            }
        },
        moveUp: function () {
            this.hoveringValue = common.getPreviousOption(this.filteredData, this.hoveringValue);
            if (this.resultsElement) {
                var lastScrollTopIndex = common.getLastScrollTopIndex(this.hoveringValue, this.resultsElement, this.filteredData, this.lastScrollTopIndex);
                if (lastScrollTopIndex !== null) {
                    this.lastScrollTopIndex = lastScrollTopIndex;
                }
            }
        },
        moveDown: function () {
            this.hoveringValue = common.getNextOption(this.filteredData, this.hoveringValue);
            if (this.resultsElement) {
                var lastScrollTopIndex = common.getLastScrollTopIndex(this.hoveringValue, this.resultsElement, this.filteredData, this.lastScrollTopIndex);
                if (lastScrollTopIndex !== null) {
                    this.lastScrollTopIndex = lastScrollTopIndex;
                }
            }
        },
        selectByEnter: function () {
            if (this.hoveringValue) {
                var option = common.getOptionByValue(this.data, this.hoveringValue);
                this.select(option);
            }
        },
    }
});
